### PR TITLE
feat: introduce hanko profile element and related api changes

### PR DESCRIPTION
**Pull request by bjoern-m**
# Description

Adds the following features to the Hanko backend and frontend to match the requirements of the hanko-profile element:

- Adds an Email API. The user can create multiple email addresses, one email address per user can be marked as the primary email address and the primary email address can't be deleted. The email address provided during account creation becomes the primary email automatically (during account creation, when email verification is turned off, or when using a passkey otherwise) and can be changed later on (via `POST /emails/:id/set_primary`). 
- Introduces a new config parameter `emails.require_verification`. It's turned on by default. When turned off, email verification is skipped during account creation. Please note that the previous version of hanko-elements will currently only work properly when `emails.require_verification` is turned off, because the updated passcode API would not know where to send the passcode when it's a new user with no email address assigned. Update: @FreddyDevelop created a workaround to overcome this issue.
- A Passcode initialisation request accepts an optional parameter `email_id`, to send the passcode to a specific email address. When the parameter is not provided, the primary email address is used, so the API behaves like before the changes (as long the user has a email address assigned). The `email_id` will be stored to the passcode table, and during passcode finalisation, the specific email address will be marked as verified (and as primary). 
- Adds new endpoints to change and delete WebAuthn credentials.

# Implementation

Database Changes:

- New table `primary_emails` has been created. Ensures one user can have one primary email 
- New table `emails`  has been created, to store multiple email addresses to a user. Emails referenced by the primary email table can't be deleted.
- New column `email_id` added to `passcodes` table, so we can updated the verified status of the specific email address.
- The columns `users.email` and `users.verified` have been dropped. 
- New `name` column added to `WebAuthnCredentials`, which is just a name for the user, without any technical relevance.

API Changes:

- New route `GET /emails` to list the email addresses of the current user.
- New route `POST /emails` to add new email addresses to the current user. A new email address remains unverified until passcode verification has been succeeded. 
- New route `POST /emails/:id/set_primary` to mark a specific email as the primary email.
- New route `DELETE /emails/:id` to delete the given email address. Will return an error when it's the primary email.
- New parameter in request body for `passcode.Init()`: You can now optionally provide an `email_id` to achieve that you can send a passcode to an unassigned email address or to a specific email address that ist already assigned to a user.
- Route `PATCH /users/:id` removed from admin router, because the user now only consists of an ID, so nothing to patch is left.
- New route `GET /webauthn/credentials` returns a list of credentials assigned to the current user.
- New route `PATCH /webauthn/credentials/:id`, used to update the credential name.
- New route `DELETE /webauthn/credentials/:id` deletes a credential.
- `POST /users`, `POST /user` and `GET /user/:id` are now returning individual DTO's instead of the user model, but they stay compatible with the previous version of hanko-auth. Some properties have been marked as deprecated.

Frontend changes:
- New web component <hanko-profile> included
- Styles and CSS-variables have been updated

# API Tests

At least I mocked the persistence classes and changed the existing tests. I also created some positive tests for the new email features, but I had difficulties and no indication if I am doing it right and using the test suite as intended. Need some help, or deliver it later.
 
# SDK Tests

Tests are complete

# Documentation

The API spec, SDK docs and Hanko-elements readme have been updated.

# Todos

- Better API testing
- Update E2E tests

### CODE REVIEW

File: backend/handler/passcode.go [hunk 1]
```diff
@@ -231,11 +261,31 @@ func (h *PasscodeHandler) Finish(c echo.Context) error {
...
-				return fmt.Errorf("failed to update user: %w", err)
+				return fmt.Errorf("failed to create audit log: %w", err)
+			}
+
+			if user.Emails.GetPrimary() == nil {
+				primaryEmail := models.NewPrimaryEmail(passcode.Email.ID, user.ID)
+				err = primaryEmailPersister.Create(*primaryEmail)
+				if err != nil {
+					return fmt.Errorf("failed to create primary email: %w", err)
+				}
```
Comment (FreddyDevelop) [inline backend/handler/passcode.go:288]: I think we should also add an audit log here, that the primary email was added (or changed).

File: backend/handler/user.go [hunk 2]
```diff
@@ -44,24 +50,81 @@ func (h *UserHandler) Create(c echo.Context) error {
...
-		user, err := h.persister.GetUserPersisterWithConnection(tx).GetByEmail(body.Email)
+		newUser := models.NewUser()
+		err := h.persister.GetUserPersisterWithConnection(tx).Create(newUser)
+		if err != nil {
+			return fmt.Errorf("failed to store user: %w", err)
+		}
+
+		email, err := h.persister.GetEmailPersisterWithConnection(tx).FindByAddress(body.Email)
 		if err != nil {
 			return fmt.Errorf("failed to get user: %w", err)
```
Comment (FreddyDevelop) [inline backend/handler/user.go:61]:
```suggestion
			return fmt.Errorf("failed to get email: %w", err)
```

File: backend/handler/user.go [hunk 3]
```diff
@@ -103,23 +177,26 @@ func (h *UserHandler) GetUserIdByEmail(c echo.Context) error {
...
-		UserId                string `json:"id"`
-		Verified              bool   `json:"verified"`
-		HasWebauthnCredential bool   `json:"has_webauthn_credential"`
-	}{
-		UserId:                user.ID.String(),
-		Verified:              user.Verified,
-		HasWebauthnCredential: len(user.WebauthnCredentials) > 0,
+	credentials, err := h.persister.GetWebauthnCredentialPersister().GetFromUser(*email.UserID)
+	if err != nil {
+		return fmt.Errorf("failed to get user: %w", err)
```
Comment (FreddyDevelop) [inline backend/handler/user.go:192]:
```suggestion
		return fmt.Errorf("failed to get webauthn credentials: %w", err)
```

File: backend/handler/webauthn.go [hunk 4]
```diff
@@ -392,6 +392,124 @@ func (h *WebauthnHandler) FinishAuthentication(c echo.Context) error {
...
+		return fmt.Errorf("failed to get user: %w", err)
+	}
+
+	credential, err := h.persister.GetWebauthnCredentialPersister().Get(credentialID)
+	if err != nil {
+		return fmt.Errorf("failed to get webauthn credentials: %w", err)
+	}
+
+	if credential == nil || credential.UserId.String() != user.ID.String() {
+		return dto.NewHTTPError(http.StatusNotFound).SetInternal(errors.New("the user does not have a webauthn with the specified credentialId"))
```
Comment (FreddyDevelop) [inline backend/handler/webauthn.go:451]:
```suggestion
		return dto.NewHTTPError(http.StatusNotFound).SetInternal(errors.New("the user does not have a webauthn credential with the specified credentialId"))
```

File: backend/handler/webauthn.go [hunk 5]
```diff
@@ -392,6 +392,124 @@ func (h *WebauthnHandler) FinishAuthentication(c echo.Context) error {
...
+
+	credentialId := c.Param("id")
+
+	credential, err := h.persister.GetWebauthnCredentialPersister().Get(credentialId)
+	if err != nil {
+		return fmt.Errorf("failed to get webauthn credential: %w", err)
+	}
+
+	if credential == nil || credential.UserId.String() != user.ID.String() {
+		return dto.NewHTTPError(http.StatusNotFound).SetInternal(errors.New("webauthn credential not found"))
```
Comment (FreddyDevelop) [inline backend/handler/webauthn.go:495]:
```suggestion
		return dto.NewHTTPError(http.StatusNotFound).SetInternal(errors.New("the user does not have a webauthn credential with the specified credentialId"))
```

File: frontend/elements/src/components/form/Button.tsx [hunk 6]
```diff
@@ -45,13 +49,13 @@ const Button = ({
         secondary ? styles.secondary : styles.primary
       )}
     >
-      <LoadingIndicator
+      <LoadingSpinner
         isLoading={isLoading}
         isSuccess={isSuccess}
-        secondary={!secondary}
+        secondary={true}
       >
```
[the remark this answers is not in the document]
Comment (bjoern-m) [reply frontend/elements/src/components/form/Button.tsx:56]: Yes, the button is always disabled when in loading state and then it's the secondary loading spinner no matter if it is displayed on the primary or secondary button:

Screenshot 2023-01-20 at 12 04 31
(image shows secondary loading spinner on primary button)

[the remark this answers is not in the document]
Comment (FreddyDevelop) [reply frontend/elements/src/components/form/Button.tsx:56]: 👍

File: backend/persistence/models/audit_log.go [hunk 7]
```diff
@@ -43,4 +43,11 @@ var (
 	AuditLogWebAuthnAuthenticationInitFailed     AuditLogType = "webauthn_authentication_init_failed"
 	AuditLogWebAuthnAuthenticationFinalSucceeded AuditLogType = "webauthn_authentication_final_succeeded"
 	AuditLogWebAuthnAuthenticationFinalFailed    AuditLogType = "webauthn_authentication_final_failed"
+	AuditLogWebAuthnCredentialUpdated            AuditLogType = "webauthn_credential_updated"
+	AuditLogWebAuthnCredentialDeleted            AuditLogType = "webauthn_credential_deleted"
```
[the remark this answers is not in the document]
Comment (bjoern-m) [reply backend/persistence/models/audit_log.go:47]: @sonatype-lift ignore

File: backend/persistence/models/audit_log.go [hunk 8]
```diff
@@ -43,4 +43,11 @@ var (
 	AuditLogWebAuthnAuthenticationInitFailed     AuditLogType = "webauthn_authentication_init_failed"
 	AuditLogWebAuthnAuthenticationFinalSucceeded AuditLogType = "webauthn_authentication_final_succeeded"
 	AuditLogWebAuthnAuthenticationFinalFailed    AuditLogType = "webauthn_authentication_final_failed"
+	AuditLogWebAuthnCredentialUpdated            AuditLogType = "webauthn_credential_updated"
```
[the remark this answers is not in the document]
Comment (bjoern-m) [reply backend/persistence/models/audit_log.go:46]: @sonatype-lift ignore

File: backend/handler/user.go [hunk 9]
```diff
@@ -44,24 +50,91 @@ func (h *UserHandler) Create(c echo.Context) error {
...
-		return c.JSON(http.StatusOK, newUser)
+		// This cookie is a workaround for hanko element versions before 0.1.0-alpha,
+		// because else the backend would not know where to send the first passcode.
+		c.SetCookie(&http.Cookie{
+			Name:     "hanko_email_id",
+			Value:    email.ID.String(),
+			Domain:   h.cfg.Session.Cookie.Domain,
+			Secure:   h.cfg.Session.Cookie.Secure,
+			HttpOnly: h.cfg.Session.Cookie.HttpOnly,
+		})
```
[the remark this answers is not in the document]
Comment (bjoern-m) [reply backend/handler/user.go:131]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (bjoern-m) [reply backend/handler/user.go:131]: @sonatype-lift ignore

**Review by FreddyDevelop — APPROVED** [review]
LGTM

**Comment by like-a-bause** [comment]
I just tried the following:
* Run this PRs quickstart.
* Login with an email I already registered via passkey with a previous hanko version 
* Add a passkey in the profile
* Delete the passkey I just created
* Logout
* Try to log in again -> fails
* When logging in again via passcode the "old" passkey is still shown there

**Comment by like-a-bause** [comment]
Just talked with @FreddyDevelop about this and it's out-of-scope for this PR. Opened a new issue.

**Review by like-a-bause — APPROVED** [review]

**Pull request merged**